### ISSUE TITLE
Version annotation should not use default value for storage true

### ISSUE THREAD

**Issue by iocanel** [fabric8io/kubernetes-client#2806]
The storage attribute needs to be present once and only once between different versions of a CRD.
Having the default value set to `true` is counter intuitive, as then users would have to go around and explicitly set false to all other versions.
This makes for a really bad user experience.

In other words storage is something that needs to be explicitly set, by the user and not assumed by default.

**Comment by metacosm** [comment]
That means that the default use case where there's only one version will be more cumbersome, though.

**Comment by iocanel** [comment]
> That means that the default use case where there's only one version will be more cumbersome, though.

That's true, however the moment that you start adding a second version it starts feeling un-natural if not annoying:
`Having to go through the non-storage versions and disable them in order to define which is the actual storage version`.

Now, to lift the burden of the default use case, we can internally set `storage=true` when there is only a single version. If this is documented and clearly stated in the javadoc, I don't see any reason why that would be a problem.

**Comment by metacosm** [comment]
I feel the current way is more natural: you start with one version. It's stored. You add a mew version: you usually deactivate the old one and store the new one so you only need to deactivate the old one. When you add yet another version, presumably you still want to store that new version so you only need to deactivate the previous one.

The only case where things are a little annoying is when you need to migrate legacy code with multiple versions at once. Otherwise the current behavior makes sense for greenfield projects, which I think is what we should optimize for.

**Comment by metacosm** [comment]
We should take a decision on this so that it's addressed for 5.2.

**Comment by manusa** [comment]
Is there a way we can detect if there is more than one version of a CRD and then act accordingly (if > 1 version: no default storage, if only 1 version use default storage + if no default storage selected log warning)

**Comment by metacosm** [comment]
That feels like too magical. The rules need to be simple, imo, and people need to make explicit decisions on which version is stored so I really feel it's natural to go through your existing versions and make a choice when you add a new version.

**Comment by metacosm** [comment]
Can we close this issue, @iocanel ?

**Comment by andreaTP** [comment]
> However, I'd still point out that a warning would be nice in case there's more than one version and more than one active storage (previous versions weren't deactivated).

Kubernetes will not accept the generated CRD.

Overall I agree that it will be nice to have at least one of:
 - an explicit `storage` version setting
 - a warning in case the resulting CRD is not going to be valid (containing multiple versions with `storage = true`)

**Comment by andreaTP** [comment]
@metacosm @iocanel @manusa I think there is still time and value to act on this issue, wdyt?
